libc/src/time/ctime_s.cpp

```diff
@@ -0,0 +1,49 @@
+//===-- Implementation of ctime_s function --------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "ctime_s.h"
+#include "hdr/errno_macros.h"
+#include "src/__support/CPP/limits.h"
+#include "src/__support/common.h"
+#include "src/__support/macros/config.h"
+#include "time_utils.h"
+
+namespace LIBC_NAMESPACE_DECL {
+
+using LIBC_NAMESPACE::time_utils::TimeConstants;
+
+LLVM_LIBC_FUNCTION(int, ctime_s,
+                   (char *buffer, rsize_t buffer_size, const time_t *t_ptr)) {
+  // TODO (https://github.com/llvm/llvm-project/issues/115907): invoke
+  // constraint handler
+  if (buffer == nullptr || t_ptr == nullptr)
+    return EINVAL;
+
+  if (*t_ptr > cpp::numeric_limits<int32_t>::max()) {
+    if (buffer_size != 0 && buffer_size <= RSIZE_MAX)
```

Comment:
> So the setting of buffer[0] to 0 only needs to occur for the rsize bounds check, not the time numerics limits check. Caveat being if I understood the standard correctly.

Seems likely, confirm me on that.

Regarding the zero check, I actually have done exactly what you said. Only thing that this condition is here since I added the bounds `[26 - RSIZE_MAX]` later so it was needed here.

Reply:
Ah! Sorry, I missed that `ASCTIME_MAX_BYTES` was the check for 26. I guess if you re-ordered that check to occur _before_ the time max check, then you could skip checking buffer size again on L28.

i.e.
```c++
  if (buffer_size < TimeConstants::ASCTIME_MAX_BYTES ||
      buffer_size > RSIZE_MAX) {
    buffer[0] = '\0';
    return ERANGE;
  }

  if (*t_ptr > cpp::numeric_limits<int32_t>::max())
    return EINVAL;
```

Reply:
I don't think we need to set `buffer[0]` in any other case; that should be removed below (L42).